src/lib.rs

```diff
@@ -248,31 +258,307 @@ impl<T, const N: usize> ArrayVec<T, { N }> {
+    /// 3 previously mentioned conditions yourself before calling this
+    /// method.
+    #[inline]
+    pub unsafe fn insert_unchecked(
+        &mut self,
+        index: usize,
+        item: T,
+        len: usize,
+    ) {
+        self.insert_unchecked_keep_len(index, item, len);
+        self.set_len(len + 1);
+    }
+
+    /// Insert an item into the vector without checking if the index is
+    /// valid or if the vector isn't full or the vector's length and
+    /// without incrementing the vector's length.
+    ///
+    /// # Safety
+    ///
+    /// If you plan on using this function, you need to check for the
+    /// 3 previously mentioned conditions yourself before calling this
+    /// method. You also need to increment the vector's length afterward
+    /// yourself.
+    pub unsafe fn insert_unchecked_keep_len(
+        &mut self,
+        index: usize,
+        item: T,
+        len: usize,
+    ) {
+        // The spot to put the Vec::new(new value);
+        let p = self.as_mut_ptr().add(index);
+        // Shift everything over to make space. (Duplicating the
+        // `index`th element into two consecutive places.)
+        ptr::copy(p, p.add(1), len - index);
+        // Write it in, overwriting the first copy of the `index`th
+        // element.
+        ptr::write(p, item);
+    }
+
+    /// Remove the value contained at `index` and return it.
+    ///
+    /// # Panics
+    ///
+    /// The index is out of bounds.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use const_arrayvec::ArrayVec;
+    /// let mut vector = ArrayVec::from([4, 3, 2]);
+    ///
+    /// let three = vector.remove(1);
+    ///
+    /// assert_eq!(three, 3);
+    /// assert_eq!(&vector, [4, 2].as_ref());
+    /// ```
+    pub fn remove(&mut self, index: usize) -> T {
+        if let Some(item) = self.try_remove(index) {
+            item
+        } else {
+            out_of_bounds!("remove", index, self.len());
+        }
+    }
+
+    /// If `index` is in bounds, remove the value contained at `index`
+    /// and return it.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use const_arrayvec::ArrayVec;
+    /// let mut vector = ArrayVec::from([4, 3, 2]);
+    ///
+    /// let three = vector.try_remove(1);
+    /// let what = vector.try_remove(24);
+    ///
+    /// assert_eq!(three, Some(3));
+    /// assert_eq!(what, None);
+    /// assert_eq!(&vector, [4, 2].as_ref());
+    /// ```
+    #[inline]
+    pub fn try_remove(&mut self, index: usize) -> Option<T> {
+        if index < self.len() {
+            Some(unsafe { self.remove_unchecked(index) })
+        } else {
+            None
+        }
+    }
+
+    /// Remove the value contained at `index` and return it.
+    ///
+    /// # Safety
+    ///
+    /// The index must be in bounds.
+    pub unsafe fn remove_unchecked(&mut self, index: usize) -> T {
+        let len = self.len();
+
+        // Where the value to remove is.
```

Comment:
Thanks for leaving lots of comments about what you're doing and the assumptions made :+1:

Reply:
No problems! It also helps me while proof-reading my code.